src/test/java/com/flipkart/zjsonpatch/TestNodesEmissionTest.java

```diff
@@ -0,0 +1,94 @@
+package com.flipkart.zjsonpatch;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.IOException;
+import java.util.EnumSet;
+
+import static org.junit.Assert.*;
+import org.junit.Test;
+
+public class TestNodesEmissionTest {
+
+    private static ObjectMapper mapper = new ObjectMapper();
+
+    private static EnumSet<DiffFlags> flags;
+
+    static {
+        flags = DiffFlags.defaults();
+        flags.add(DiffFlags.EMIT_TEST_OPERATIONS);
+    }
+
+    @Test
+    public void testNodeEmittedBeforeReplaceOperation() throws IOException {
+        JsonNode source = mapper.readTree("{\"key\":\"original\"}");
+        JsonNode target = mapper.readTree("{\"key\":\"replaced\"}");
+
+        JsonNode diff = JsonDiff.asJson(source, target, flags);
+
+        JsonNode testNode = mapper.readTree("{\"op\":\"test\",\"path\":\"/key\",\"value\":\"original\"}");
+        assertEquals(2, diff.size());
+        assertEquals(testNode, diff.iterator().next());
+    }
```

Comment:
It looks very good! Thank you!

But I could also suggest extends tests not only check `test` node present on patch, but it also correctly handled on `apply`  operation by complimentary add apply test for each current test like:
```java
    @Test
    public void testNodeEmittedBeforeReplaceOperation() throws IOException {
        JsonNode source = mapper.readTree("{\"key\":\"original\"}");
        JsonNode target = mapper.readTree("{\"key\":\"replaced\"}");

        JsonNode diff = JsonDiff.asJson(source, target, flags);

        assertEquals(2, diff.size());
        Iterator<JsonNode> iterator = diff.iterator();
        assertEquals(mapper.readTree("{\"op\":\"test\",\"path\":\"/key\",\"value\":\"original\"}"), iterator.next());
        assertEquals(mapper.readTree("{\"op\":\"replace\",\"path\":\"/key\",\"value\":\"replaced\"}"), iterator.next());
    }

    @Rule
    public ExpectedException exceptionRule = ExpectedException.none();

    @Test
    public void testApplyWithTestOnReplaceOperation() throws IOException {
        exceptionRule.expect(JsonPatchApplicationException.class);
        exceptionRule.expectMessage("[TEST Operation] value mismatch");

        JsonNode source = mapper.readTree("{\"key\":\"original\"}");
        JsonNode target = mapper.readTree("{\"key\":\"replaced\"}");

        JsonNode diff = JsonDiff.asJson(source, target, flags);

        source = mapper.readTree("{\"key\":\"changed\"}"); // Change source
        JsonPatch.apply(diff, source);
    }
```

It will be also very good to see context at what it failed, as I suggest in issue. So instead of just message `[TEST Operation] value mismatch` provide path and values, both expected and present. What are you think?

Reply:
IMO that's covered by separate tests (the ones dealing with `TEST` ops). As for the error message, I'd treat that as a separate change request, though I certainly agree that more information is better...